Index: glib-2.20.0/docs/reference/glib/glib-docs.sgml
===================================================================
--- glib-2.20.0.orig/docs/reference/glib/glib-docs.sgml
+++ glib-2.20.0/docs/reference/glib/glib-docs.sgml
@@ -241,6 +241,8 @@ synchronize their operation.
   </index>  
   <index role="2.20">
     <title>Index of new symbols in 2.20</title>
-  </index>  
-
+  </index> 
+  <index role="maemo">
+    <title>Index of new symbols in Maemo</title>
+  </index> 
 </book>
Index: glib-2.20.0/docs/reference/glib/glib-sections.txt
===================================================================
--- glib-2.20.0.orig/docs/reference/glib/glib-sections.txt
+++ glib-2.20.0/docs/reference/glib/glib-sections.txt
@@ -2465,6 +2465,7 @@ g_unichar_break_type
 g_unichar_combining_class
 g_unicode_canonical_ordering
 g_unicode_canonical_decomposition
+g_unicode_canonical_decomposition_to_buffer
 g_unichar_get_mirror_char
 GUnicodeScript
 g_unichar_get_script
