diff --git a/configure.in b/configure.in
index eaa6943..e3264cd 100644
--- a/configure.in
+++ b/configure.in
@@ -2522,6 +2522,9 @@ if test x"$GCC" = xyes; then
         dnl AC_DEFINE_UNQUOTED(G_ATOMIC_ARM_LINUX, 1,
         dnl                   [special arm linux implementation])
         dnl glib_memory_barrier_needed=yes
+        # some ARM CPUs and gcc versions need flags for proper backtraces
+        ARM_FRAME_FLAGS="-fno-omit-frame-pointer -mapcs-frame"
+        AC_SUBST(ARM_FRAME_FLAGS)
         ;;
       crisv32*|etraxfs*)
         AC_MSG_RESULT([crisv32])
diff --git a/glib/Makefile.am b/glib/Makefile.am
index 97eae45..8b4e5dd 100644
--- a/glib/Makefile.am
+++ b/glib/Makefile.am
@@ -38,6 +38,9 @@ AM_CPPFLAGS = 				\
 	-DGLIB_COMPILATION 		\
 	-DPCRE_STATIC
 
+# force frame-pointers for source files with noreturn functions for backtraces
+AM_CFLAGS = $(patsubst %, $(ARM_FRAME_FLAGS), $(filter $(<F), gmessages.c gtestutils.c ))
+
 glib.def: glib.symbols
 	(echo -e EXPORTS; $(CPP) -P -DINCLUDE_VARIABLES -DINCLUDE_INTERNAL_SYMBOLS -DG_OS_WIN32 -DALL_FILES - <$(srcdir)/glib.symbols | sed -e '/^$$/d' -e 's/^/	/' -e 's/G_GNUC_[^ ]*//g') > glib.def
 
