Index: glib-2.20.3/configure.in
===================================================================
--- glib-2.20.3.orig/configure.in	2009-07-22 13:15:42.000000000 +0300
+++ glib-2.20.3/configure.in	2009-07-22 13:16:08.000000000 +0300
@@ -2410,6 +2410,9 @@
       dnl AC_DEFINE_UNQUOTED(G_ATOMIC_ARM_LINUX, 1,
       dnl                   [special arm linux implementation])
       dnl glib_memory_barrier_needed=yes
+      # some ARM CPUs and gcc versions need flags for proper backtraces
+      ARM_FRAME_FLAGS="-fno-omit-frame-pointer -mapcs-frame"
+      AC_SUBST(ARM_FRAME_FLAGS)
       ;;
     crisv32*|etraxfs*)
       AC_MSG_RESULT([crisv32])
Index: glib-2.20.3/glib/Makefile.am
===================================================================
--- glib-2.20.3.orig/glib/Makefile.am	2009-07-22 13:15:29.000000000 +0300
+++ glib-2.20.3/glib/Makefile.am	2009-07-22 13:15:42.000000000 +0300
@@ -32,6 +32,9 @@
 	-DGLIB_COMPILATION 		\
 	-DPCRE_STATIC
 
+# force frame-pointers for source files with noreturn functions for backtraces
+AM_CFLAGS = $(patsubst %, $(ARM_FRAME_FLAGS), $(filter $(<F), gmessages.c gtestutils.c ))
+
 glib.def: glib.symbols
 	(echo -e EXPORTS; $(CPP) -P -DINCLUDE_VARIABLES -DINCLUDE_INTERNAL_SYMBOLS -DG_OS_WIN32 -DALL_FILES - <$(srcdir)/glib.symbols | sed -e '/^$$/d' -e 's/^/	/' -e 's/G_GNUC_[^ ]*//g') > glib.def
 
