Index: glib-2.20.1/glib/gfileutils.c
===================================================================
--- glib-2.20.1.orig/glib/gfileutils.c	2009-05-14 15:13:04.000000000 +0200
+++ glib-2.20.1/glib/gfileutils.c	2009-05-14 15:17:33.000000000 +0200
@@ -987,7 +987,14 @@ write_to_temp_file (const gchar  *conten
 #endif
   
   errno = 0;
+#ifdef MAEMO_CHANGES
+  /* introduce a write-barrier for file contents before close()/rename()
+   * to work around file systems with writeback behaviour like UBIFS.
+   */
+  if (fdatasync (fileno (file)) < 0 || fclose (file) == EOF)
+#else /* !MAEMO_CHANGES */
   if (fclose (file) == EOF)
+#endif /* MAEMO_CHANGES */
     { 
       save_errno = errno;
       
