diff --git a/gio/xdgmime/xdgmime.c b/gio/xdgmime/xdgmime.c
index 0888583..c5c76ed 100644
--- a/gio/xdgmime/xdgmime.c
+++ b/gio/xdgmime/xdgmime.c
@@ -418,6 +418,8 @@ xdg_check_dirs (void)
 static int
 xdg_check_time_and_dirs (void)
 {
+  static gboolean check_interval_checked = FALSE;
+  static gint check_interval = 5;
   struct timeval tv;
   time_t current_time;
   int retval = FALSE;
@@ -425,7 +427,18 @@ xdg_check_time_and_dirs (void)
   gettimeofday (&tv, NULL);
   current_time = tv.tv_sec;
 
-  if (current_time >= last_stat_time + 5)
+  if (!check_interval_checked)
+    {
+      const char *env;
+
+      env = getenv ("XDG_CHECK_INTERVAL");
+      if (env != NULL)
+        check_interval = atoi(env);
+      check_interval_checked = TRUE;
+    }
+
+  if (current_time >= last_stat_time + check_interval
+      && (!last_stat_time || check_interval > 0))
     {
       retval = xdg_check_dirs ();
       last_stat_time = current_time;
